### PR TITLE
[ADD] base_web_hook: Create abstract web hooks

### PR DESCRIPTION
**Pull request by lasley**
* Create a module to allow for the receipt of web hooks in an abstract fashion

Depends:
- [ ] #3 

This is a PoC.

### CODE REVIEW

File: base_web_hook/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import http
+
+
+class WebHookController(http.Controller):
+
+    @http.route(
+        ['/base_web_hook/<string:slug>'],
+        type='json',
```
Comment (lasley) [inline base_web_hook/controllers/main.py]: Hmmmm maybe should be an HTTP receiver as well

**Comment by lasley** [comment]
Alright this is implemented in a connector, so it's less rough around the edges now. We've still got some testing to do on the connector, and another one to implement, so I would assume there's still work to come here.

[This is an implementation example](https://github.com/LasLabs/odoo-connector-helpscout/pull/3)

File: base_web_hook/models/web_hook.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+except ImportError:
+    _logger.info('`python-slugify` Python library not installed.')
+
+
+class WebHook(models.Model):
+
+    _name = 'web.hook'
+    _description = 'Web Hook'
+
+    name = fields.Char(
```
Comment (lasley) [inline base_web_hook/models/web_hook.py:26]: Add `UNIQUE(name)`

**Comment by lasley** [comment]
Alright I've now added a Request Bin hook type, which simply catches and saves inbound requests. This is a useful tool for testing web hooks as an administrator, but will also help me test the module in an automated fashion.

**Comment by hbrunn** [comment]
late to the party, are you aware of https://github.com/OCA/server-tools/tree/10.0/webhook ?
(I do have a couple of things I'd like to change...)

**Comment by lasley** [comment]
Not aware at all. How did you conquer the receipt of JSON?

**Comment by hbrunn** [comment]
I didn't conquer anything there, didn't even use the module. But I read it yesterday and decided I can't use it for what I had in mind.

**Comment by hbrunn** [comment]
looks straightforward enough: https://github.com/OCA/server-tools/blob/10.0/webhook/controllers/main.py#L17

**Comment by lasley** [comment]
I'm looking at the module, and it definitely does not work as intended in most cases. The route is using `type='json'`, which means that the hook must be an RPC formatted request. See https://github.com/odoo/odoo/issues/7766

I would propose this module as a direct successor to the other TBH. There are significantly more features here, such as the ability to receive non-JSON requests & session authenticated hooks.

**Comment by lasley** [comment]
@moylop260 - your module https://github.com/OCA/server-tools/tree/10.0/webhook has instructions specifically for Github web hooks, but odoo/odoo#7766 alludes to the fact that this does not actual work. Can you confirm OCA/server-tools/webhook does work for any inbound JSON webhook without sitting a proxy in front of the Odoo server?

File: base_json_request/http.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+old_handle_exception = http.JsonRequest._handle_exception
+old_init = http.JsonRequest.__init__
+
+
+def __init__(self, *args):
+    try:
+        old_init(self, *args)
+    except BadRequest as e:
+        try:
+            args = self.httprequest.args
```
Comment (pntabuye) [inline base_json_request/http.py:21]: AttributeError: 'JsonRequest' object has no attribute 'httprequest'

**Review by pntabuye — COMMENTED** [review]
tested on Odoo 10e and not conclusive. The non JSON-RPC json does not raise a BadRequest error from werkzeug.

File: base_json_request/README.rst [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+=================
+
+This module allows you to receive JSON requests in Odoo that are not
+RPC.
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/210/10.0
+
+Bug Tracker
```
Comment (mlaitinen) [inline base_json_request/README.rst:16]: Please add a usage section on how to use this module

File: base_web_hook/README.rst [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/210/10.0
+
+Known Issues
+============
+
+* Security is too lax; public can read too much. Maybe should also add a group for hooks..
+* Buffer length should be checked before ``httprequest.get_data`` calls
+
+Bug Tracker
```
Comment (mlaitinen) [inline base_web_hook/README.rst:21]: Please add a usage section on how to use this module

**Review by mlaitinen — CHANGES_REQUESTED** [review]

**Comment by gurneyalex** [comment]
sale, old odoo running on python 2 -> closing

**Pull request closed without merging**